### PR TITLE
If building with -Oz, do not unroll memcpy(), memset() and memmove() 

### PR DESCRIPTION
**Pull request by juj**
Here is an experiment with routing `memcpy()`, `memset()` and `memmove()` to JS side in `-Oz` builds.

Using the following code as size test:
`a.c`
```c
#include <emscripten.h>
#include <emscripten/em_math.h>
#include <string.h>
/*
void *memset(void *str, int c, size_t n)
{
	unsigned char *s = (unsigned char *)str;
	while(n--) *s++ = c;
	return str;
}

void *memcpy(void *dest, const void *src, size_t n)
{
	unsigned char *d = (unsigned char *)dest;
	const unsigned char *s = (const unsigned char *)src;
	while(n--) *d++ = *s++;
	return dest;
}

void *memmove(void *dest, const void *src, size_t n)
{
	if (dest < src) return memcpy(dest, src, n);
	unsigned char *d = (unsigned char *)dest + n;
	const unsigned char *s = (const unsigned char *)src + n;
	while(n--) *--d = *--s;
	return dest;
}
*/

int a[8] = { 0, 1, 2, 3, 4, 5, 6, 7 };
int b[8];

int main(int argc, char **argv)
{
	memcpy(b, a, emscripten_math_random() >= 0 ? 8*4 : 0);
//	EM_ASM(console.log(`b: ${$0}, ${$1}, ${$2}, ${$3}, ${$4}, ${$5}, ${$6}, ${$7}`), b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]);

	memset(b, 0x7F, emscripten_math_random() >= 0 ? 8*4 : 0);
//	EM_ASM(console.log(`b: ${$0.toString(16)}, ${$1.toString(16)}, ${$2.toString(16)}, ${$3.toString(16)}, ${$4.toString(16)}, ${$5.toString(16)}, ${$6.toString(16)}, ${$7.toString(16)}`), b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]);

	memmove(a, a+1, emscripten_math_random() >= 0 ? 7*4 : 0);
//	EM_ASM(console.log(`a: ${$0}, ${$1}, ${$2}, ${$3}, ${$4}, ${$5}, ${$6}, ${$7}`), a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7]);

	memmove(a+1, a, emscripten_math_random() >= 0 ? 7*4 : 0);
//	EM_ASM(console.log(`a: ${$0}, ${$1}, ${$2}, ${$3}, ${$4}, ${$5}, ${$6}, ${$7}`), a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7]);
}
```

First, building with
```
emcc a.c -o a.html -Oz -s MINIMAL_RUNTIME=2 --closure 1 -s SUPPORT_ERRNO=0 -s TEXTDECODER=2 -s ENVIRONMENT=web -mbulk-memory 
```
gives a total html+js+wasm size of 669+173+224=1066 bytes for the output. In `-mbulk-memory`, `memory.fill` and `memory.copy` are used instead of these JS intrinsics.

Then, building with
```
emcc a.c -o a.html -Oz -s MINIMAL_RUNTIME=2 --closure 1 -s SUPPORT_ERRNO=0 -s TEXTDECODER=2 -s ENVIRONMENT=web
```
gives a total size of 669+313+235=1217 bytes. +151 bytes/+14.1% more than bulk memory.

Then, uncommenting the inline C implementations of memcpy/memset/memove in the test file above (to override JS side implementations) gives a total size of 669+173+390=1232 bytes. +15 bytes more than routing to JS side.

Building the test code against master branch for a baseline outside this PR, and with the original unrolled emscripten_memcpy/memset/memmove implementations gives 669+236+1551 bytes=2456 bytes. +1239 bytes/+101.8% more than the JS routed code.

Best code size and performance will come with `-mbulk-memory`, although its browser marketshare is not yet ubiquitous.

Not sure if we should land something like this. It does help minimal sized code in the absence of `-mbulk-memory`.

Unfortunately I have to stop this investigation short here, don't have more time to look into this atm. Still wanted to post this since this came up in the other convesation.

The other side of this, i.e. performance, is something that should be analyzed before landing this. Even though `-Oz` is "minimize all code size", it is still an open question whether the +15 bytes larger Wasm side implementations might make more sense than routing to JS, since there may be an unrealistic performance cliff to some applications if they route thousands of memset/memcpy calls out to JS, as opposed to staying inside Wasm.

In the absence of perf benchmarks, I'd be open to landing this under a `-s JS_MEMOPS=1` flag (instead of gating on -Oz).

### CODE REVIEW

**Comment by kripken** [comment]
Another option here is de-optimize memcpy/memset/etc. The current impl has manual unrolling, handling of misaligned cases, etc. A simple loop on single bytes might be comparable in size to calling out to a JS call. It would be slower on large memcopies, though.

**Comment by juj** [comment]
> Another option here is de-optimize memcpy/memset/etc.

Yeah, that is the implementation in `a.c` that was benchmarked. I think that would correspond to what the deoptimized code would look like.

**Comment by juj** [comment]
Adjusted the PR to remove loop unrolling. I think this would be good to land in this form.

**Review by kripken — APPROVED** [review]
I'm happy with this. @sbc100 what do you think?

**Comment by sbc100** [comment]
I see to these are no longer JS functions?   can you update the PR title and description?

File: src/library.js
Comment (sbc100) [inline src/library.js:723]: Should this be SHRINK_LEVEL > 2 ?    I would have thought higher SHRINK_LEVEL would mean more compression.

Is there any advantage to doing this?   Won't this only be included on demand anyway?

File: tools/system_libs.py
Comment (sbc100) [inline tools/system_libs.py:541]: How about "-tiny" or "-small"?

File: tools/system_libs.py
Comment (sbc100) [inline tools/system_libs.py:533]: Can we come up with more better name?   How about just `SizeOptimizedLibrary`?

File: system/lib/libc/emscripten_memset.c
Comment (sbc100) [inline system/lib/libc/emscripten_memset.c:12]: nit: We are trying to move towards a single coding style based on the LLVM style (see .clang-format at the top level).. which means braces on the same line as the decl.

**Review by sbc100 — COMMENTED** [review]
Nice!   I live that we ended up with a native solution (so that STANDALONE_WASM didn't have to be different.

**Review by sbc100 — APPROVED** [review]

File: src/library.js
Comment (juj) [reply src/library.js:726]: No, in `SHRINK_LEVEL >= 2`, this function is not used, but instead the Wasm side function is used. (we never call out to JS in `SHRINK_LEVEL >= 2`)

File: tools/system_libs.py
Comment (juj) [reply tools/system_libs.py:542]: I don't think that is a better name, since it loses the meaning of what the corresponding `-O?` flag is. If I read `-tiny` or `-small`, I won't immediately know ("oh, do I get `-tiny` if I do `-Os`, or do I have to do pass `-Oz` for that?)

File: tools/system_libs.py
Comment (juj) [reply tools/system_libs.py:534]: I think for the same reason as above, that would not be as good. Maybe `OptimizedForOzLibrary` if something? I want the name reflect via static reading that it is for `-Oz` and not for `-Os`

File: system/lib/libc/emscripten_memset.c
Comment (juj) [reply system/lib/libc/emscripten_memset.c]: Not a fan of changing code style on the grounds of LLVM, but if it helps, I don't mind.

Comment (sbc100) [reply system/lib/libc/emscripten_memset.c]: Are you not a fan of the LLVM style or not a fan of having a style in general?

IIRC I chose LLVM because it what we use in upstream LLVM (obviously) and also in binaryen and similar to the chromium style used in wabt.   So it just what most of the team is use to.

I'm up for discussing what style we should use. but I don't like he idea of not having any consistent style at all.  I also like the idea of using clang-format to avoid extended style discussions and making it super easy to conform to the style without adding work... so I'm hoping we can do that here too.

File: src/library.js
Comment (sbc100) [reply src/library.js:726]: So I'm a little confused by the comment then .. it says `In -Oz builds, we replace memcpy() altogether with a JS variant,` but that contractions "we never call out to JS in SHRINK_LEVEL >= 2"

Should the comment say `we replace memcpy() altogether with a native variant` ?

File: tools/system_libs.py
Comment (sbc100) [reply tools/system_libs.py:542]: OK that make sense.

**Review by sbc100 — APPROVED** [review]

File: system/lib/libc/emscripten_memset.c
Comment (juj) [reply system/lib/libc/emscripten_memset.c]: Not a fan of the { on decl line, I don't think we followed that for .c code before.

The general issue I have with styling is that it has lead to all sorts of checkers - clang-format and flake and js-lint tools added to the repo. They are making contributing from Windows (and on macOS, because of the MacPorts vs Homebrew vs whatever you have installed manually vs macOS version - conflict hell) more of a barrier to entry than it used to be. The tooling may be practically seamless for Linux, but on Windows and on macOS one needs to hunt down installation of tools that are not there by default.

Clang-format on Visual Studio is particularly annoying: it hangs VS often, it formats right in the middle of you typing, it repeatedly brings in popups "hey we notice that you have a clang-format in this repo so we just reformatted for you", so that using VS to edit any files in Emscripten repo has become a dont-do-that thing.

That is why I see that these kinds of styling requirements generally add more work than is worth. I do wish I could also use Linux all day and not care about Windows and macOS, but that is where our users are, hence needing to make sure things roll fine on those OSes.

**Comment by juj** [comment]
> So I'm a little confused by the comment then .. it says In -Oz builds, we replace memcpy() altogether with a JS variant, but that contractions "we never call out to JS in SHRINK_LEVEL >= 2"

> Should the comment say we replace memcpy() altogether with a native variant ?

Good catch! The comment was out of date, updated.

File: system/lib/libc/emscripten_memset.c
Comment (sbc100) [reply system/lib/libc/emscripten_memset.c]: I see you point.  If some developers find it hard to run these tools then I agree we should not make it hard requirement in this repo.

However, I don't think that means we should not have a preferred style for the code in the repro or that we should not make a best effort to conform to it where possible/easy.

Comment (sbc100) [reply system/lib/libc/emscripten_memset.c]: Per the above I'm not blocking this from landing based on style here.  I leave it up to you.

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
PR description looks like it needs updating ("to JS library functions" seems wrong).

File: src/library.js
Comment (juj) [reply src/library.js:726]: Good catch, the comment was outdated - fixed now.

**Pull request merged**
